src/python/pants/backend/python/tasks/python_bundle.py

```diff
@@ -0,0 +1,86 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import os
+
+from pants.backend.python.targets.python_app import PythonApp
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TargetDefinitionException
+from pants.fs import archive
+from pants.task.task import Task
+from pants.util import dirutil
+
+
+class PythonBundle(Task):
+  """Create an archive bundle of PythonApp targets."""
+
+  _DEPLOYABLE_ARCHIVES = 'deployable_archives'
+  _PEX_ARCHIVES = 'pex_archives'
+
+  @classmethod
+  def product_types(cls):
+    return [cls._DEPLOYABLE_ARCHIVES]
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(PythonBundle, cls).prepare(options, round_manager)
+    round_manager.require_data(cls._PEX_ARCHIVES)
+
+  def __init__(self, *args, **kwargs):
+    super(PythonBundle, self).__init__(*args, **kwargs)
+    self._outdir = self.get_options().pants_distdir
+
+  def execute(self):
+    targets_to_bundle = self.context.targets(PythonApp.is_python_app)
+    dirutil.safe_mkdir(self._outdir)
+    for target in targets_to_bundle:
+      bundle_dir = os.path.join(self.workdir, target.id)
```

Comment:
Rather than doing this manually, would recommend setting the `self.create_target_dirs` property. Example:
https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/jvm/tasks/provide_tools_jar.py#L38-L41

Then you can use `vt.results_dir` as the unique directory for a target.

Reply:
I concur, but also, you're not invalidating here, so you have no `vt` at all... See the jvm backend's `bundle_create.py` for an example of how to only act on invalid targets (but still set up products and populate `dist/` for all targets).

Reply:
Updated to use `create_target_dirs`.

Currently the logic is pretty similar, though there are slight differences between JVM/Python bundle logic. This is definitely an area for refactoring in the future. At this time I tried to be minimally invasive with the JVM bundle path.